ruler-gradle-plugin/src/main/kotlin/com/spotify/ruler/plugin/RulerTask.kt

```diff
@@ -76,24 +77,61 @@ abstract class RulerTask : DefaultTask() {
         val dependencies = getDependencies() // Get all entries from all dependencies
 
         // Attribute bundle entries and group into components
-        val attributor = Attributor(DependencyComponent(project.path, ComponentType.INTERNAL))
-        val components = attributor.attribute(files, dependencies)
+        val components = attributeAndGroupEntries(files, dependencies)
 
         val ownershipInfo = getOwnershipInfo() // Get ownership information for all components
         generateReports(components, ownershipInfo)
     }
 
-    private fun getFilesFromBundle(): List<AppFile> {
+    private fun attributeAndGroupEntries(
+        files: Map<String, List<AppFile>>,
+        dependencies: Map<String, List<DependencyComponent>>
+    ): Map<DependencyComponent, List<AppFile>> {
+        val components = hashMapOf<DependencyComponent, List<AppFile>>()
+        files.forEach { (projectPath, f) ->
+            val attributor = Attributor(
+                DependencyComponent(
+                    projectPath, if (projectPath == project.path) {
+                        ComponentType.INTERNAL
+                    } else {
+                        ComponentType.DYNAMIC
+                    }
+                )
+            )
+            components += attributor.attribute(f, dependencies)
+        }
+        return components
+    }
+
+    private fun getFilesFromBundle(): Map<String, List<AppFile>> {
         val apkCreator = ApkCreator(project.rootDir)
-        val splits = apkCreator.createSplitApks(bundleFile.asFile.get(), deviceSpec.get(), workingDir.asFile.get())
+        val splits = apkCreator.createSplitApks(
+            bundleFile.asFile.get(),
+            deviceSpec.get(),
+            workingDir.asFile.get()
+        )
         val apks = splits.listFiles() ?: emptyArray()
-
         val apkParser = ApkParser()
-        val entries = apks.flatMap(apkParser::parse)
-
         val classNameSanitizer = ClassNameSanitizer(mappingFile.asFile.orNull)
         val apkSanitizer = ApkSanitizer(classNameSanitizer)
-        return apkSanitizer.sanitize(entries)
+
+        // get project path to app file list
+        val files = hashMapOf<String, List<AppFile>>()
+        val dynamicFeatureModules = getDynamicFeatureModules()
+        (dynamicFeatureModules.keys + "base").forEach { dfm ->
+            val entries = apks.filter { it.name.startsWith(dfm) }.flatMap(apkParser::parse)
```

Comment:
I hadn't thought of `:sample:dynamicfeature:foo` before. This situation does need to be considered. I'll try to find a better way.

Reply:
Can't find any API to obtain artifacts of DFM.
`project.project(":sample:dynamicfeature").configurations.getByName("${appInfo.variantName}RuntimeClasspath").incoming` contains artifacts of `:app` because DMF implements `:app`😂

Reply:
The case of having different versions of an artifact in the base app and the DFM is very curious, I would expect that the conflict would be resolved and only one version loaded 🤔 

Seems like it's tricky to do what we're trying to do, maybe we can go for a simpler solution instead. I.e. we could treat DFMs totally separate from the base app. In the JSON that could looks something like this:
```json
{
  "name": "com.spotify.music",
  "dynamicFeatureModules": [
    {
      "name": "dynamicfeature",
      "downloadSize": 234,    
      "installSize": 123,
      "owner": "foo",
      "files": []
    }
  ]
}
```

We could simply extract that information from the different APK splits. And in the HTML we could have a separate `Dynamic` tab that lists those modules like the components are listed in the `Breakdown` tab. Seems like that would probably be simple enough to implement and also provide enough information to users to be useful.

Reply:
> The case of having different versions of an artifact in the base app and the DFM is very curious, I would expect that the conflict would be resolved and only one version loaded 🤔
> 
> Seems like it's tricky to do what we're trying to do, maybe we can go for a simpler solution instead. I.e. we could treat DFMs totally separate from the base app. In the JSON that could looks something like this:
> 
> ```json
> {
>   "name": "com.spotify.music",
>   "dynamicFeatureModules": [
>     {
>       "name": "dynamicfeature",
>       "downloadSize": 234,    
>       "installSize": 123,
>       "owner": "foo",
>       "files": []
>     }
>   ]
> }
> ```
> 
> We could simply extract that information from the different APK splits. And in the HTML we could have a separate `Dynamic` tab that lists those modules like the components are listed in the `Breakdown` tab. Seems like that would probably be simple enough to implement and also provide enough information to users to be useful.

@simonschiller Refer to [ApkSetUtils](https://github.com/google/bundletool/blob/96669880398d3df3fc8cfe12c82f3e94fa7ef36e/src/test/java/com/android/tools/build/bundletool/testing/ApkSetUtils.java#L48), I have modified code, only support json report for now. Please review.